runtime/lua/tohtml.lua

```diff
@@ -0,0 +1,1263 @@
+          end
+        end
+        for _, v in ipairs(cell[3]) do
+          if v[2] then
+            s = s .. (name_to_tag(state.highlights_name[v[2]]))
+          end
+          s = s .. (html_escape(v[1], state.tabstop))
+          if v[2] then
+            s = s .. (name_to_closetag(state.highlights_name[v[2]]))
+          end
+        end
+        char = cell[4][#cell[4]]
+      end
+      if col == #line + 1 and not char then
+        break
+      end
+      if hide_count == 0 then
+        char = char
+          or { vim.api.nvim_buf_get_text(state.bufnr, row - 1, col - 1, row - 1, col, {})[1] }
+        if char[2] then
+          s = s .. name_to_tag(state.highlights_name[char[2]])
+        end
+        s = s .. html_escape(char[1], state.tabstop)
+        if char[2] then
+          s = s .. name_to_closetag(state.highlights_name[char[2]])
+        end
+      end
+    end
+    table.insert(out, s)
+  end
+  for row = 1, state.buflen + 1 do
+    loop(row)
+  end
+  assert(#stack == 0, 'an open html tag was never closed')
+  table.insert(out, '</pre>')
+end
+
+--- @param out string[]
+--- @param fn fun()
+local function extend_body(out, fn)
+  table.insert(out, '<body style="display: flex">')
+  fn()
+  table.insert(out, '</body>')
+end
+
+--- @param out string[]
+--- @param fn fun()
+local function extend_html(out, fn)
+  table.insert(out, '<!DOCTYPE html>')
+  table.insert(out, '<html>')
+  fn()
+  table.insert(out, '</html>')
+end
+
+--- @param winnr number
+--- @param global_state vim.tohtml.state.global
+--- @return vim.tohtml.state
+local function global_state_to_state(winnr, global_state)
+  local bufnr = vim.api.nvim_win_get_buf(winnr)
+  local opt = global_state.conf
+  local width = opt.width or vim.bo[bufnr].textwidth
+  if not width or width < 1 then
+    width = vim.api.nvim_win_get_width(winnr)
+  end
+  local state = setmetatable({
+    winnr = winnr == 0 and vim.api.nvim_get_current_win() or winnr,
+    opt = vim.wo[winnr],
+    style = generate_styletable(bufnr),
+    bufnr = bufnr,
+    tabstop = (' '):rep(vim.bo[bufnr].tabstop),
+    width = width,
+    buflen = vim.api.nvim_buf_line_count(bufnr),
+  }, { __index = global_state })
+  return state --[[@as vim.tohtml.state]]
+end
+
+--- @param opt vim.tohtml.opt
+--- @param title? string
+--- @return vim.tohtml.state.global
+local function opt_to_global_state(opt, title)
+  local fonts = {}
+  if opt.font then
+    fonts = type(opt.font) == 'string' and { opt.font } or opt.font --[[@as (string[])]]
+  elseif vim.o.guifont:match('^[^:]+') then
+    table.insert(fonts, vim.o.guifont:match('^[^:]+'))
+  end
+  table.insert(fonts, 'monospace')
+  --- @type vim.tohtml.state.global
+  local state = {
+    background = get_background_color(),
+    foreground = get_foreground_color(),
+    title = opt.title or title or false,
+    font = table.concat(fonts, ','),
+    highlights_name = {},
+    conf = opt,
+  }
+  return state
+end
+
+--- @type function[]
```

Comment:
```suggestion
--- @type fun(state: vim.tohtml.state)[]
```

It looks like all these functions have the same signature.